packages/attune/src/server/repo/index/sign.rs

```diff
@@ -116,29 +92,76 @@ pub async fn handler(
+    // transaction commits and _before_ the index update in S3 completes. This
+    // will leave the repository in an inconsistent state, but can be fixed by
+    // redoing the S3 upload from database state.
+    //
+    // There's also a small risk that two changes A and B will commit in order
+    // `A -> B`, but will upload indexes to S3 in order `B -> A`, which will
+    // cause the older index to overwrite the newer one. This is also very
+    // unlikely, but there is no good mitigation here besides a cron job. Note
+    // that any _subsequent_ upload will still upload the correct indexes,
+    // because the _database_ state is transactionally consistent.
+    apply_change_to_s3(&state.s3, &repo, &req, &result, previous_by_hash_indexes).await;
+
+    Ok(Json(SignIndexResponse {}))
+}
+
+async fn apply_change_to_db(
+    tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    tenant_id: &TenantID,
+    req: &SignIndexRequest,
+) -> Result<(PackageChangeResult, Option<PreviousByHashIndexes>), ErrorResponse> {
+    // Verify the request cleartext signature.
+    let (public_key, _headers) = SignedPublicKey::from_string(&req.public_key_cert)
+        .expect("could not parse public key certificate");
+    debug!(?public_key, "public key");
+    if let Err(e) = public_key.verify() {
+        return Err(ErrorResponse::new(
+            StatusCode::BAD_REQUEST,
+            "PUBLIC_KEY_VERIFICATION_FAILED".to_string(),
+            format!("could not verify public key: {e}"),
+        ));
+    }
+    let (clearsigned, _headers) = CleartextSignedMessage::from_string(&req.clearsigned)
+        .expect("could not parse clearsigned index");
+    debug!(clearsigned = ?clearsigned.text(), "clearsigned index");
+    if let Err(e) = clearsigned.verify(&public_key) {
+        return Err(ErrorResponse::new(
+            StatusCode::BAD_REQUEST,
+            "CLEARSIGN_VERIFICATION_FAILED".to_string(),
+            format!("could not verify clearsigned index: {e}"),
+        ));
+    }
 
     // Replay the diff onto the current state of the index. Since index
     // generation is deterministic, this should yield the same index that was
     // signed locally.
     let result =
-        generate_release_file_with_change(&mut tx, &tenant_id, &req.change, req.release_ts).await?;
+        generate_release_file_with_change(tx, tenant_id, &req.change, req.release_ts).await?;
     debug!(?result, "replayed index");
 
     // Compare the replayed index with the signed index.
     // If the signatures match, this validates that the index signed by the client is the same as the one we replayed.
     let (detachsigned, _headers) = StandaloneSignature::from_string(&req.detachsigned)
         .expect("could not parse detached signature");
-    tracing::debug!(index = ?result.release_file.contents, ?detachsigned, "detachsigned index");
+    debug!(index = ?result.release_file.contents, ?detachsigned, "detachsigned index");
```

Comment:
i tend to write a lot of stuff like this with the fully qualified path. would you rather i not do this?